no errors but a big mess with merge conflicts and no experience

ArneTillmann opened the pull request:
Ich hatte große Schwierigkeiten mit Git, aber die sollten jetzt aus dem Weg geräumt sein. Ich würde mich freuen, wenn du einmal drüber schauen könntest und vielleicht mal einen kurzen Kommentar abgeben könntest. Ich will die Sachen noch auskommentieren die nächsten Tage und mir deine Email im Detail anschauen und die Vorschläge daraus umsetzten. 
Achso es muss noch ein paar Rechenfehler geben, weil aus irgendeinem Grund der eine Wert kleiner als 1 ist...

ArneTillmann committed:
no errors but a big mess with merge conflicts and no experience

ArneTillmann committed:
Revised with use off pep8. May still not be fully consistend though

simeoncarstens commented on comparison.py:
```diff
@@ -1,29 +1,64 @@
-    return -0.5* np.sum(x**2 )
-#gradient der log Dichte funktion
+
+    return -1./2 * np.sum(x**2)
```
```suggestion
    return -0.5 * np.sum(x**2)
```
I usually try to keep formulas in code as easy to read as possible. This and the following two suggestions are what I would write, but it's probably quite subjective.

simeoncarstens commented on comparison.py:
```diff
@@ -1,29 +1,64 @@
 def gradient_log_prob(x):
-    return -1*x
+
+    return -1 * x
```
```suggestion
    return -x
```

simeoncarstens commented on comparison.py:
```diff
@@ -1,29 +1,64 @@
-leapfrog_HMC = LeapfrogHMC(x_0, chain_length, log_prob, gradient_log_prob, stepsize, trajectory_length)
-chain, acceptance_rate = leapfrog_HMC.build_chain()
-print(chain[:10])
+    return -1 * np.identity(2)
```
```suggestion
    return -np.identity(2)
```

simeoncarstens commented on comparison.py:
```diff
@@ -1,29 +1,64 @@
+
+"""creating the chains for the different integrators respectively"""
+u7_HMCs = [U7HMC(log_prob, gradient_log_prob, hessian_log_prog,
+                 stepsize, trajectory_length) for i in range(number_of_chains)]
```
```suggestion
u7_HMCs = [U7HMC(log_prob, gradient_log_prob, hessian_log_prog,
                 stepsize, trajectory_length) for _ in range(number_of_chains)]
```
In Python, it's somewhat standard practice to denote unneeded variables with and underscore `_` so as to not make the reader wonder what, in this case, the `i` might be for.

simeoncarstens commented on comparison.py:
```diff
@@ -1,29 +1,64 @@
+
+for i in range(number_of_chains):
+    chain, acceptance_rate = u7_HMCs[i].build_chain(x_0, chain_length)
+    chainu7_HMCs.append(chain)
```
```suggestion
for sampler in u7_HMCs:
    chain, acceptance_rate = sampler.build_chain(x_0, chain_length)
    chainu7_HMCs.append(chain)
```
If possible, it's always better do iterate directly over the objects you're eventually using instead of iterating over a range of indices and then accessing objects by index, which is error-prone and less readable.

simeoncarstens commented on comparison.py:
```diff
@@ -1,29 +1,64 @@
+
+Leapfrog_HMCs = [LeapfrogHMC(log_prob, gradient_log_prob, stepsize,
+                             trajectory_length)
+                             for i in range(number_of_chains)]
```
See above ( use `_`). Also, variable names should start with a lower-case letter (`leapfrog_HMCs`).

simeoncarstens commented on comparison.py:
```diff
@@ -1,29 +1,64 @@
+
+for i in range(number_of_chains):
+    chain, acceptance_rate = Leapfrog_HMCs[i].build_chain(x_0, chain_length)
+    chainleapfrog_HMCs.append(chain)
```
See above (iterate over `Leapfrog_HMCs`)

simeoncarstens commented on hmc_sampler.py:
```diff
@@ -1,47 +1,66 @@
+        trajectory length for the integrator simulating the Hamiltonian dynamics
+
+        IMPORTANT!!!
+        I am using the gradient of the density function and not
```
```suggestion
        I am using the gradient of the log-density function and not
```

simeoncarstens commented on comparison.py:
```diff
@@ -1,29 +1,64 @@
+"""Evaluation of the inte by using the revised Gelman-Rubin Diagnostic
+see https://arxiv.org/pdf/1812.09384.pdf for further notice
+"""
+b = int(np.floor(chain_length**(1.0/3)))
+a = int(np.floor(chain_length/b))
```
```suggestion
b = int(np.floor(chain_length ** (1.0 / 3)))
a = int(np.floor(chain_length / b))
```
According to PEP8, it's recommended to have spaces around operators such as `**`, `/` and relations like `>`, `==`, etc.

simeoncarstens commented on gelman_rubin.py:
```diff
@@ -1,44 +1,73 @@
+"""This is the implementation of the revised Gelman-Rubin Diagnostic
+with use of the presented calculations given in the paper
+https://arxiv.org/pdf/1812.09384.pdf
+written by Dootika Vats and Christina Knudson
+"""
```
```suggestion
"""This is the implementation of the revised Gelman-Rubin Diagnostic
with use of the presented calculations given in the paper
https://arxiv.org/pdf/1812.09384.pdf
written by Dootika Vats and Christina Knudson
"""
import numpy as np
```
Module doc strings always go to the very top of the file, before `import` statements.

simeoncarstens commented on gelman_rubin.py:
```diff
@@ -1,44 +1,73 @@
+
+
+def improved_estimator_PSRF(b, a, m, n, p, chains):
+    """The final revised gelman-rubin statistic receiving b,a as additional
+    parameters for the calculation of the k_th_batch_mean.
+    They are chosen in the way, that n=a*b, whereas n is the chain length
+    (number of samples in one chain). Usual choices for b include round off
+    n^(1/2) or round off n^(1/3) according to the authers.
+    m is the number of chains, n their length, p the dimension of the target
+    distribution and chains the np.ndarray containing the chains in order [m,n,p]
+    """
```
```suggestion
    """The final revised Gelman-Rubin statistic receiving b, a as additional
    parameters for the calculation of the k_th_batch_mean.
    They are chosen in the way, that n=a*b. Usual choices for b include round off
    n^(1/2) or round off n^(1/3) according to the authors.
    
    Arguments:
    - m (int): number of chains
    - n (int): length of each chain
    - p (int): dimension of the target distribution 
    - chains (numpy.ndarray): array containing the chains in order [m,n,p]
    """
```
In doc strings, it's often good practice to write an easily readable list of the function arguments, which also indicates their types. There are [standards](https://stackoverflow.com/a/24385103/1656472) for that (my suggestions above conforms to none), such that docstrings can be parsed by software such as [Sphinx](https://www.sphinx-doc.org/en/master/) to automatically generate nicely-formatted documentation.

simeoncarstens reviewed (changes requested):
Only a first pass to point out some coding style issues.
Good job on the formatting! Looks much, much more readable now. In the Gelman-Rubin code, you should take care to consistently put spaces around math operators such as `/`, `**`. I also usually try to avoid unnecessary fractions - instead of `1. / 2` I would write `0.5`, which is more readable, I think. Same for exponents: `bla ** (1. / 2)` is much less readable compared to a simple `sqrt(bla)`, but if it's to agree with the typesetting of formulas in a reference paper, you might have a point.
Your docstrings are good; they contain most of the information and context a reader needs. But I nevertheless gave an example of an improved docstring.
I'll check the implementation of the Gelman-Rubin statistic and will propose more changes in a second review, in case I stumble on a bug.

ArneTillmann committed:
Update comparison.py

Well, it might be subjective to some extend. Anyway, I agree with you and will change it.

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Update comparison.py

Ok, this is an obvious one...

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Update comparison.py

same

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Update comparison.py

Thats a good one. I'll keep that in mind

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Update comparison.py

True, also really good hint. THX or kiitos (finish)

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Update hmc_sampler.py

yup

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Update comparison.py

Ups, I must have gotten that the wrong way, but thank you.

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Update gelman_rubin.py

Alright ^^

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Update gelman_rubin.py

Sounds very reasonable, I had a look at the HMC classes as well and did the same there

Co-authored-by: Simeon Carstens <simeon.carstens@tweag.io>

ArneTillmann committed:
Style ajustments

ArneTillmann committed:
Merge branch 'Gelman_Rubin' of https://github.com/ArneTillmann/HMC_Integrator_Comparison into Gelman_Rubin

ArneTillmann committed:
small style and test updates

ArneTillmann committed:
gitignore and random pos semi def cov matrix

ArneTillmann committed:
safe

ArneTillmann committed:
quite nice result for two dim

ArneTillmann committed:
example for dim 10 tuned

ArneTillmann committed:
vergessen
